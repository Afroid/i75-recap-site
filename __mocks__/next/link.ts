import * as React from "react";
import type { ReactNode } from "react";
import type { UrlObject } from "url";
import type { AnchorHTMLAttributes } from "react";

interface LinkMockProps
  // drop the built-in href so we can replace its type
  extends Omit<AnchorHTMLAttributes<HTMLAnchorElement>, "href">
{
  href: string | UrlObject;
  children: ReactNode;
}

export default function LinkMock({
  href,
  children,
  ...rest
}: LinkMockProps) {
  let url = "";

  if (typeof href === "string") {
    url = href;
  } else {
    // coalesce out both `undefined` and `null`:
    const pathname = href.pathname ?? "";
    url = pathname;

    const { query } = href;
    if (query && typeof query === "object") {
      const params = new URLSearchParams(
        query as Record<string, string>
      ).toString();
      if (params) {
        url += "?" + params;
      }
    }
  }

  return React.createElement(
    "a",
    { href: url, ...rest },
    children
  );
}
