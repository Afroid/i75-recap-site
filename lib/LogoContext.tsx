"use client";
import { createContext, useContext, useMemo } from "react";

const logos = [
  "/default-logo.png",
  "/default-logo-alt.png",
  "/default-RWB.png",
];

const LogoContext = createContext<string | null>(null);

export function LogoProvider({ children }: { children: React.ReactNode }) {
  // Randomizes only once on initial render
  const logo = useMemo(
    () => logos[Math.floor(Math.random() * logos.length)],
    []
  );
  return (
    <LogoContext.Provider value={logo}>
      {children}
    </LogoContext.Provider>
  );
}

export function useLogo() {
  const logo = useContext(LogoContext);
  if (logo === null) throw new Error("useLogo must be inside LogoProvider");
  return logo;
}
