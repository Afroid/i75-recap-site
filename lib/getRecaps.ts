import recap2024 from "@/data/recaps/2024.json";
import recap2023 from "@/data/recaps/2023.json";
import { RecapData } from "@/types/types";

export function getRecaps(): Record<number, number[]> {
  const allData: RecapData[] = [recap2024, recap2023]; // Add more as needed
  const summary: Record<number, number[]> = {};

  for (const season of allData) {
    summary[season.year] = season.recaps.map((r) => r.week);
  }

  return summary;
}
