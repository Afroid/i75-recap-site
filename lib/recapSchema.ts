import { z } from "zod";

const RecapMatchupSchema = z.object({
  team1: z.string(),
  team2: z.string(),
  winner: z.string(),
  score: z.string(),
  breakdown: z.string(),
  imageUrl: z.string().optional(),
  imageAlt: z.string().optional(),
});

const GameNotesSectionSchema = z.object({
  type: z.literal("gameNotes"),
  pointsLeader: z.object({
    name: z.string(),
    points: z.number(),
    notes: z.string().optional(),
  }),
  dud: z.object({
    name: z.string(),
    points: z.number(),
    notes: z.string().optional(),
  }),
  matchups: z.array(RecapMatchupSchema),
});

const RecapSectionSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("intro"),
    content: z.string(),
  }),
  z.object({
    type: z.literal("tidbits"),
    intro: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    outro: z.string().optional(),
    imageUrl: z.string().optional(),
    imageAlt: z.string().optional(),
  }),
  z.object({
    type: z.literal("injuries"),
    intro: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    outro: z.string().optional(),
    imageUrl: z.string().optional(),
    imageAlt: z.string().optional(),
  }),
  z.object({
    type: z.literal("sidePieces"),
    intro: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    outro: z.string().optional(),
    imageUrl: z.string().optional(),
    imageAlt: z.string().optional(),
  }),
  z.object({
    type: z.literal("matchups"),
    intro: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    outro: z.string().optional(),
    imageUrl: z.string().optional(),
    imageAlt: z.string().optional(),
  }),
  z.object({
    type: z.literal("farewell"),
    intro: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    outro: z.string().optional(),
    imageUrl: z.string().optional(),
    imageAlt: z.string().optional(),
  })
]);

const RecapWeekSchema = z.object({
  week: z.number(),
  title: z.string(),
  sections: z.array(z.union([RecapSectionSchema, GameNotesSectionSchema])),
});

export const RecapDataSchema = z.object({
  year: z.number(),
  recaps: z.array(RecapWeekSchema),
});
