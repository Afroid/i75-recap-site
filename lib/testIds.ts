export const TestIds = {
  /**
   * Pages
   */

  // Index - App Level
  SITE_HEADER: "site-header",
  MAIN_CONTENT: "main-content",
  PAGE_WRAPPER: "page-wrapper",

  // Home Page
  HOME_MAIN:   "home-main",
  HOME_HERO_LOGO: "home-hero-logo",
  HOME_INTRO_SECTION: "home-intro-section",
  HOME_HERO_HEADING: "home-hero-heading",
  HOME_CTA_CONTAINER: "home-cta-container",
  HOME_VIEW_ALL_RECAPS_BUTTON: "home-view-all-recaps-button",
  HOME_LEARN_MORE_BUTTON: "home-learn-more-button",

  // 404 Page
  NOT_FOUND_MAIN: "not-found-main",
  RETURN_HOME_LINK: "return-home-link",

  // About Page
  ABOUT_MAIN: "about-main",

  // Contact Page
  CONTACT_MAIN: "contact-main",
  CONTACT_EMAIL_LINK: "contact-email-link",
  CONTACT_FORM: "contact-form",
  CONTACT_FORM_NAME_SECTION: "contact-form-name-section",
  CONTACT_FORM_EMAIL_SECTION: "contact-form-email-section",
  CONTACT_FORM_MESSAGE_SECTION: "contact-form-message-section",
  CONTACT_FORM_SUBMIT: "contact-form-submit",

  /**
   * Recap Pages - (pages/recaps)
   */

  // Recap Page

  // View All Recaps Page
  VIEW_ALL_RECAPS: "view-all-recaps",
  // Year Page

  /**
   * Components
   */

  // Breadcrumb Component
  BREADCRUMB: "breadcrumb",
  BREADCRUMB_NAV: "breadcrumb-nav",
  BREADCRUMB_LIST: "breadcrumb-list",
  BREADCRUMB_ITEM: "breadcrumb-item",
  BREADRCRUMB_LINK: "breadcrumb-link",
  BREADCRUMB_CURRENT: "breadcrumb-current",
  BREADCRUMB_SEPARATOR: "breadcrumb-separator",

  // DesktopNavbar Component

  // GameNotesSection.tsx
  GAME_NOTES_SECTION: "game-notes-section",
  GAME_NOTES_ITEM: "game-notes-item",
  GAME_NOTES_SCOREBOARD: "game-notes-scoreboard",
  GAME_NOTES_BREAKDOWN: "game-notes-breakdown",
  GAME_NOTES_IMAGE_CONTAINER: "game-notes-image-container",
  GAME_NOTES_IMAGE: "game-notes-image",

  // Header.tsx
  HEADER_SITE_HEADER: "site-header",
  HEADER_MOBILE_HEADER: "mobile-header",
  HEADER_HAMBURGER_BUTTON: "header-hamburger-button",
  HEADER_MOBILE_LOGO: "header-mobile-logo",
  HEADER_DESKTOP_LOGO: "header-desktop-logo",
  HEADER_DESKTOP_NAV: "header-desktop-nav",
  HEADER_MOBILE_DRAWER: "header-mobile-drawer",

  // HeaderLogo.tsx
  HEADER_LOGO_LINK: "header-logo-link",
  HEADER_LOGO_BG_DARK: "header-logo-bg-dark",
  HEADER_LOGO_BG_LIGHT: "header-logo-bg-light",
  HEADER_LOGO_CONTAINER: "header-logo-container",
  HEADER_LOGO_NAME: "header-logo-name",
  HEADER_LOGO_DYNAMIC: "header-logo-dynamic",

  // LeaderDudSection.tsx
  LDS_SECTION: "lds-section",
  LDS_LEADER_HEADING: "lds-leader-heading",
  LDS_LEADER_POINTS: "lds-leader-points",
  LDS_LEADER_NOTES: "lds-leader-notes",
  LDS_DUD_HEADING: "lds-dud-heading",
  LDS_DUD_POINTS: "lds-dud-points",
  LDS_DUD_NOTES: "lds-dud-notes",

  // MobileDrawer.tsx
  MOBILE_SEASON_TOGGLE: "mobile-season-toggle",
  MOBILE_SEASON_WEEKS: "mobile-season-weeks",
  MOBILE_WEEK_LINK: "mobile-week-link",
  MOBILE_DRAWER_OVERLAY: "mobile-drawer-overlay",
  MOBILE_DRAWER: "mobile-drawer",
  MOBILE_DRAWER_TITLE: "mobile-drawer-title",
  MOBILE_OLDER_SEASON_TITLE: "mobile-older-season-title",
  MOBILE_TOGGLE_OLDER_SEASONS: "mobile-toggle-older-seasons",

  // RecapLayout.tsx
  RECAP_TITLE: "recap-title",
  RECAP_SECTION: "recap-section",
  RECAP_HEADING: "recap-heading",
  RECAP_INTRO: "recap-intro",
  RECAP_BULLETS: "recap-bullets",
  RECAP_BULLET: "recap-bullet",
  RECAP_CONTENT: "recap-content",
  RECAP_IMAGE: "recap-image",

  // Scoreboard.tsx
  SCOREBOARD_CONTAINER: "scoreboard-container",
  SCOREBOARD_ENTRY: "scoreboard-entry",
  SCOREBOARD_LOGO: "scoreboard-logo",
  SCOREBOARD_NAME: "scoreboard-name",
  SCOREBOARD_RECORD: "scoreboard-record",
  SCOREBOARD_SCORE: "scoreboard-score",

  // TeamLogo.tsx
  TEAM_LOGO_IMAGE: "team-logo-image",

  // Static Links
  NAV_LINK_HOME: "nav-link-home",
  NAV_LINK_ABOUT: "nav-link-about",
  NAV_LINK_CONTACT: "nav-link-contact"
}
