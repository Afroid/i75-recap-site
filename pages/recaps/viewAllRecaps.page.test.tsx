jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import ViewAllRecapsPage, { getStaticProps } from "./viewAllRecaps.page";
import { getRecaps } from "@/lib/getRecaps";
import { useRouter } from "next/router";
import { TestIds } from "@/lib/testIds";

// Stub framer-motion’s <motion.li> to return a normal <li>
jest.mock("framer-motion", () => ({
  motion: {
    li: ({ children, ...props }: React.HTMLAttributes<HTMLLIElement>) => (
      <li {...props}>{children}</li>
    ),
  },
}));

jest.mock("@/lib/getRecaps");

const replaceMock = jest.fn();

beforeEach(() => {
  (useRouter as jest.Mock).mockReturnValue({
    query:    {},
    pathname: "/recaps/viewAllRecaps",
    replace:  replaceMock,
  });
  replaceMock.mockClear();
  (getRecaps as jest.Mock).mockClear();
});

describe("ViewAllRecapsPage", () => {
  const recaps = [
    { year: "2023", week: 1 },
    { year: "2022", week: 2 },
  ];

  it("renders search, dropdowns, and recap links correctly", () => {
    render(<ViewAllRecapsPage allRecaps={recaps} />);

    // Search input & icon
    const search = screen.getByPlaceholderText("Search by year or week...");
    expect(search).toBeInTheDocument();
    expect(screen.getByTestId(TestIds.VIEW_ALL_RECAPS_SEARCH_ICON)).toBeInTheDocument();

    // Year dropdown should show “All Years” label
    expect(screen.getByText("All Years")).toBeInTheDocument();

    // Sort dropdown default value is “Latest to Oldest”
    expect(screen.getByText("Latest to Oldest")).toBeInTheDocument();

    // The two recap items and links
    recaps.forEach(({ year, week }) => {
      const item = screen.getByTestId(`${TestIds.VIEW_ALL_RECAPS_RECAP_ITEM}-${year}-${week}`);
      expect(item).toBeInTheDocument();

      const link = screen.getByTestId(`${TestIds.VIEW_ALL_RECAPS_RECAP_LINK}-${year}-${week}`);

      // Assertions
      expect(link).toHaveAttribute("href", `/recaps/${year}/week-${week}`);
      expect(link).toHaveTextContent(`Week ${week}`);
    });
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<ViewAllRecapsPage allRecaps={recaps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("getStaticProps", () => {
  it("flattens getRecaps summary into allRecaps", async () => {
    // Stub getRecaps()
    (getRecaps as jest.Mock).mockReturnValue({
      "2023": [1, 2],
      "2021": [3],
    });

    // Act
    const result = await getStaticProps();

    // Assertions
    expect(getRecaps).toHaveBeenCalled();
    expect(result).toEqual({
      props: {
        allRecaps: [
          { year: "2021", week: 3 },
          { year: "2023", week: 1 },
          { year: "2023", week: 2 },
        ],
      },
    });
  });
});

describe("search behavior", () => {
  const recaps = [{ year: "2023", week: 1 }];

  beforeEach(() => {
    render(<ViewAllRecapsPage allRecaps={recaps} />);
  });

  it("calls router.replace when user types non-numeric search", () => {
    const input = screen.getByPlaceholderText("Search by year or week...");
    fireEvent.change(input, { target: { value: "foo" } });

    // Assertions
    expect(replaceMock).toHaveBeenCalledTimes(1);
    expect(replaceMock).toHaveBeenCalledWith(
      { pathname: "/recaps/viewAllRecaps", query: { search: "foo" } },
      undefined,
      { shallow: true }
    );
  });

  it("also clears year filter when typing a 4-digit year", () => {
    const input = screen.getByPlaceholderText("Search by year or week...");
    fireEvent.change(input, { target: { value: "2023" } });

    // First replace is for search=2023, second is for search=2023 & year=""
    expect(replaceMock).toHaveBeenCalledTimes(2);

    // The last call should still only send search=2023 (the empty year is stripped)
    expect(replaceMock).toHaveBeenLastCalledWith(
      { pathname: "/recaps/viewAllRecaps", query: { search: "2023" } },
      undefined,
      { shallow: true }
    );
  });
});
