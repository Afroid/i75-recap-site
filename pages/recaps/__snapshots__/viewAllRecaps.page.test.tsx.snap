// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewAllRecapsPage matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="pt-4 max-w-4xl mx-auto px-4"
    data-testid="view-all-recaps-page"
  >
    <div
      class="mb-6 flex flex-col gap-4 sm:flex-row sm:items-center sm:justify-between"
    >
      <div
        class="relative w-full sm:max-w-xs"
      >
        <input
          class="border px-3 py-2 rounded-md w-full sm:max-w-xs"
          data-testid="view-all-recaps-search-input"
          placeholder="Search by year or week..."
          type="text"
          value=""
        />
        <svg
          aria-hidden="true"
          class="lucide lucide-search absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-400"
          data-testid="view-all-recaps-search-icon"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
      </div>
      <div
        class="flex flex-col sm:flex-row w-full sm:w-auto sm:gap-4 gap-2"
      >
        <div
          class="flex flex-row gap-2 w-full sm:w-auto"
        >
          <div
            class="w-2/5 sm:w-auto lg:w-28"
          >
            <div
              class="relative w-full"
              data-testid="view-all-recaps-dropdown-
      all-years"
            >
              <button
                class="flex justify-between items-center border px-3 py-2 rounded-md w-full"
                data-testid="view-all-recaps-dropdown-button-
        all-years"
              >
                <span>
                  All Years
                </span>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-chevron-down text-gray-500"
                  data-testid="view-all-recaps-dropdown-icon-down"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m6 9 6 6 6-6"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="w-3/5 sm:w-auto lg:w-40"
          >
            <div
              class="relative w-full"
              data-testid="view-all-recaps-dropdown-
      sort-order"
            >
              <button
                class="flex justify-between items-center border px-3 py-2 rounded-md w-full"
                data-testid="view-all-recaps-dropdown-button-
        sort-order"
              >
                <span>
                  Latest to Oldest
                </span>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-chevron-down text-gray-500"
                  data-testid="view-all-recaps-dropdown-icon-down"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m6 9 6 6 6-6"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <ul
      class="grid grid-cols-1 sm:grid-cols-2 gap-6"
      data-testid="view-all-recaps-recap-list"
    >
      <li
        animate="[object Object]"
        data-testid="view-all-recaps-recap-item-2023-1"
        initial="[object Object]"
        transition="[object Object]"
      >
        <a
          class="block rounded-lg shadow-md bg-white hover:bg-green-50 p-6 text-center transition border border-gray-200 hover:border-green-400 transform transition-transform hover:shadow-lg hover:scale-105 hover:-translate-y-1 flex flex-col justify-center min-h-[150px]"
          data-testid="view-all-recaps-recap-link-2023-1"
          href="/recaps/2023/week-1"
        >
          <h2
            class="text-xl font-semibold text-gray-900 mb-2"
            data-testid="view-all-recaps-recap-title-2023-1"
          >
            Week 1
          </h2>
          <p
            class="text-gray-600"
            data-testid="view-all-recaps-recap-season-2023-1"
          >
            Season 2023
          </p>
        </a>
      </li>
      <li
        animate="[object Object]"
        data-testid="view-all-recaps-recap-item-2022-2"
        initial="[object Object]"
        transition="[object Object]"
      >
        <a
          class="block rounded-lg shadow-md bg-white hover:bg-green-50 p-6 text-center transition border border-gray-200 hover:border-green-400 transform transition-transform hover:shadow-lg hover:scale-105 hover:-translate-y-1 flex flex-col justify-center min-h-[150px]"
          data-testid="view-all-recaps-recap-link-2022-2"
          href="/recaps/2022/week-2"
        >
          <h2
            class="text-xl font-semibold text-gray-900 mb-2"
            data-testid="view-all-recaps-recap-title-2022-2"
          >
            Week 2
          </h2>
          <p
            class="text-gray-600"
            data-testid="view-all-recaps-recap-season-2022-2"
          >
            Season 2022
          </p>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
